### ISSUE TITLE
Wayland support

### ISSUE THREAD

**Issue by pieterdd** [olback/user-idle-rs#8]
Nice crate, works great on my Windows install.

On Ubuntu, Wayland appears to be the default display server now. As user-idle's README states, unfortunately there is no Wayland support yet. I assume the following bug is a consequence of this. I'll leave a log of the symptoms here so people can easily diagnose the issue through Google.

Given this program:

```rs
use user_idle::UserIdle;

fn main() {
    let idle = UserIdle::get_time().unwrap();
    println!("Idle time is {}!", idle.as_seconds());
}
```

I get the following output at runtime:

> Xlib:  extension "MIT-SCREEN-SAVER" missing on display ":0".
> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { cause: "Status not OK" }', src/main.rs:4:37
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

I ran `echo $XDG_SESSION_TYPE` to confirm that I'm on Wayland. I know this is a volunteer-ran project, but if you ever have the time and energy for it, Wayland support would be cool to have. I'm making a small cross-platform Workrave-like app so I'd like to rely on user-idle's idle detection.

**Comment by amjadorfali** [comment]
I can propose a solution to support wayland, please let me know if I should work on it @olback

**Comment by pieterdd** [comment]
Since https://github.com/olback/user-idle-rs/pull/17 remained without a response, I've published a fork with Wayland support at https://crates.io/crates/user-idle2.

**Comment by amjadorfali** [comment]
@pieterdd That is pretty awesome!

**Comment by sitompul** [comment]
@pieterdd will you add support for major DE, kde cinnamon via dbus?

* KDE through `org.kde.KIdleTime.getIdleTime`
* Cinnamon/Mint through `org.cinnamon.ScreenSaver.GetSessionIdleTime`

**Comment by pieterdd** [comment]
It's not at the top of my priority list, but it'd be nice to cover all major DEs. Steps would be:

- Install both DEs
- Reverse engineer the data structure the calls use
- Update the code
- Validate the mechanism

With X11 slowly going the way of the dodo it makes sense to go all in on D-Bus.

**Comment by amjadorfali** [comment]
@pieterdd @sitompul using a solution i stumbled upon from a fellow friend, I believe we can create Wayland support using the standard wayland APIs and without using any dbus implementation. I already used it in my project and it works for now, I will do some more testing and will probably publish another fork of Use-idle or maybe contribute to your fork!


[Implementation I'm referring to](https://github.com/amjadorfali/gtfu/blob/main/src/idle_detection/wayland_idle.rs)

[It uses the idle notification protocol for wayland](https://wayland.app/protocols/ext-idle-notify-v1)

**Comment by pieterdd** [comment]
Interesting! A solution that works independent of the user's preferred desktop environment would be nice.

**Comment by sitompul** [comment]
@amjadorfali 

https://wayland.app/protocols/ext-idle-notify-v1

Are you referring to ext_idle_notify api? 

Be sure to check it on gnome, since on the wayland page it doesn't seem to support mutter. Or you may want to add trivial dbus support on mutter/gnome.

**Comment by amjadorfali** [comment]
> @amjadorfali 
> 
> https://wayland.app/protocols/ext-idle-notify-v1
> 
> Are you referring to ext_idle_notify api? 
> 
> Be sure to check it on gnome, since on the wayland page it doesn't seem to support mutter. Or you may want to add trivial dbus support on mutter/gnome.

Yes im referring to this protocol. Tbh I most likely won't support any implementation other than this protocol to keep things simple. The ecosystem will catch up eventually, or someone might open a PR to do it. But i do think this approach is good. What do you think? I'm interested in your thoughts @sitompul @pieterdd

**Comment by sitompul** [comment]
@amjadorfali 

Any native protocol implementation will be preferred. So lgtm but fallback also needed, since mutter is the largest demographic in terms of wayland compositor

**Comment by pieterdd** [comment]
Haven't really had the time to dive into this, but it's still on my list. I'll need to familiarize myself with Wayland's ecosystem a bit.

**Comment by pieterdd** [comment]
I just noticed this:

![Image](https://github.com/user-attachments/assets/546db9da-9c2a-41f5-95b6-1a283f43fd00)

It seems that GNOME's Wayland compositor doesn't support it?

**Comment by amjadorfali** [comment]
@pieterdd yep!
There are alternate methods like using d-bus. To be completely honest I care more about stability than reach, I'm not trying to sell it I'm trying to use it 😅
In any case, I will still need some time (probably months) to start on this, I'm a little swamped these days.
